Svc/FrameAccumulator/FrameDetector/FprimeFrameDetector.cpp

```diff
@@ -0,0 +1,102 @@
+    // A better implementation would be to have CircularBuffer implement a shared interface with SerializeBufferBase,
+    // and then we could pass the CircularBuffer directly into the FrameHeader/FrameTrailer deserializers. This is left
+    // as a TODO for future improvement as it is a significant refactor
+
+    FprimeProtocol::FrameHeader header;
+    FprimeProtocol::FrameTrailer trailer;
+
+    // ---------------- Frame Header ----------------
+    // Copy CircularBuffer data into linear buffer, for serialization into FrameHeader object
+    U8 header_data[FprimeProtocol::FrameHeader::SERIALIZED_SIZE];
+    Fw::SerializeStatus status = data.peek(header_data, FprimeProtocol::FrameHeader::SERIALIZED_SIZE, 0);
+    if (status != Fw::FW_SERIALIZE_OK) {
+        return Status::NO_FRAME_DETECTED;
+    }
+    Fw::ExternalSerializeBuffer header_ser_buffer(header_data, FprimeProtocol::FrameHeader::SERIALIZED_SIZE);
+    status = header_ser_buffer.setBuffLen(FprimeProtocol::FrameHeader::SERIALIZED_SIZE);
+    FW_ASSERT(status == Fw::FW_SERIALIZE_OK, static_cast<FwAssertArgType>(status));
+    // Attempt to deserialize data into the FrameHeader object
+    status = header.deserialize(header_ser_buffer);
+    if (status != Fw::FW_SERIALIZE_OK) {
+        return Status::NO_FRAME_DETECTED;
+    }
+    // Check that deserialized start_word token matches expected value (default start_word value in the FPP object)
+    FprimeProtocol::FrameHeader default_value;
+    if (header.getstart_word() != default_value.getstart_word()) {
+        return Status::NO_FRAME_DETECTED;
+    }
+    // If the deserialized length token can't fit in current allocated size -> MORE_DATA_NEEDED
+    if (data.get_allocated_size() < FprimeProtocol::FrameHeader::SERIALIZED_SIZE + header.getlength() +
+                                        FprimeProtocol::FrameTrailer::SERIALIZED_SIZE) {
+        size_out = header.getlength() + FprimeProtocol::FrameHeader::SERIALIZED_SIZE +
+                   FprimeProtocol::FrameTrailer::SERIALIZED_SIZE;
+        return Status::MORE_DATA_NEEDED;
+    }
+
+    // ---------------- Frame Trailer ----------------
+    U8 trailer_data[FprimeProtocol::FrameTrailer::SERIALIZED_SIZE];
+    Fw::ExternalSerializeBuffer trailer_ser_buffer(trailer_data, FprimeProtocol::FrameTrailer::SERIALIZED_SIZE);
+    status = data.peek(trailer_data, FprimeProtocol::FrameTrailer::SERIALIZED_SIZE,
+                       FprimeProtocol::FrameHeader::SERIALIZED_SIZE + header.getlength());
+    if (status != Fw::FW_SERIALIZE_OK) {
+        return Status::NO_FRAME_DETECTED;
+    }
+    status = trailer_ser_buffer.setBuffLen(FprimeProtocol::FrameTrailer::SERIALIZED_SIZE);
+    FW_ASSERT(status == Fw::FW_SERIALIZE_OK, static_cast<FwAssertArgType>(status));
+    // Deserialize trailer from circular buffer (peeked data) into trailer object
+    status = trailer.deserialize(trailer_ser_buffer);
+    if (status != Fw::FW_SERIALIZE_OK) {
+        return Status::NO_FRAME_DETECTED;
```

Comment:
Should these serialization checks be asserts instead?

Reply:
No.  We don't want to fail on uplink data.